src/Microsoft.NetCore.Analyzers/Core/Security/DoNotReferSelfInSerializableClass.cs

```diff
@@ -0,0 +1,191 @@
+                                                                                                    s.Type.IsInSource());
+
+                            foreach (var field in fields)
+                            {
+                                var fieldSymbol = (ISymbol)field;
+                                var fieldType = (ISymbol)field.Type;
+
+                                AddLine(classSymbol, fieldSymbol, outDegree, forwardGraph);
+                                AddLine(fieldSymbol, fieldType, outDegree, forwardGraph);
+                                AddLine(fieldType, fieldSymbol, inDegree, invertedGraph);
+                                AddLine(fieldSymbol, classSymbol, inDegree, invertedGraph);
+                            }
+                        }, SymbolKind.NamedType);
+
+                    compilationStartAnalysisContext.RegisterCompilationEndAction(
+                        (CompilationAnalysisContext compilationAnalysisContext) =>
+                        {
+                            ModifyDegree(inDegree, forwardGraph);
+                            ModifyDegree(outDegree, invertedGraph);
+
+                            // If the degree of a vertex is greater than 0 both in the forward graph and inverted graph after topological sorting,
+                            // the vertex must belong to a loop.
+                            var leftVertices = inDegree.Where(s => s.Value > 0).Select(s => s.Key).ToImmutableHashSet();
+                            var invertedLeftVertices = outDegree.Where(s => s.Value > 0).Select(s => s.Key).ToImmutableHashSet();
+                            var verticesInLoop = leftVertices.Intersect(invertedLeftVertices);
+
+                            foreach (var vertex in verticesInLoop)
+                            {
+                                if (vertex is IFieldSymbol)
+                                {
+                                    compilationAnalysisContext.ReportDiagnostic(
+                                        vertex.CreateDiagnostic(Rule, vertex.Name));
+                                }
+                            }
+                        });
+
+                    /// <summary>
+                    /// Add a line to the graph.
+                    /// </summary>
+                    /// <param name="from">The start point of the line</param>
+                    /// <param name="to">The end point of the line</param>
+                    /// <param name="degree">The out degree of all vertices in the graph</param>
+                    /// <param name="graph">The graph</param>
+                    void AddLine(ISymbol from, ISymbol to, ConcurrentDictionary<ISymbol, int> degree, ConcurrentDictionary<ISymbol, HashSet<ISymbol>> graph)
+                    {
+                        var value = AddPoint(from, degree, graph);
+
+                        if (value.Add(to))
+                        {
+                            degree[from]++;
```

Comment:
I believe there is a data race here because `++` is not an atomic operation.

Reply:
The type of degree is ConcurrecyDictionary. And according to the API documents, the indexer's setter can store a key/value pair in the dictionary unconditionally, and overwrite the value of a key that already exists. Normally, it is used as dic[key] = dic[key] + 1. I think there's no difference between ++ operator and + 1. So, I think it can handle data race.

Reply:
I think @drewnoakes is talking about `++` (and `x = x + 1`) being three steps: read, update, and store.  The read step is atomic by itself, and the store step is atomic by itself, but "read, update, and store" all together is not atomic.

ConcurrentDictionary has an `AddOrUpdate` method which will make "read, update, and store" atomic.